summary_workflows/quantification/match_with_gt.py

```diff
@@ -0,0 +1,187 @@
+parser.add_argument('window', help='Number of bases to append to each side of the predicted site.', type=int)
+#parser.add_argument('-o', help='output file directory') # not yet implemented!!
+args = parser.parse_args()
+
+
+f_GT = args.b
+f_PD = args.a
+window = args.window
+
+
+def bedtools_window(bed1, bed2, window):
+    """
+    Python wrapper for bedtools window.
+    """
+    
+    # make sure bedtools can be called in current env
+    assert which('bedtools') is not None, "bedtools not installed or not in PATH"
+
+    # run bedtools window, capture output
+    out = subprocess.run(['bedtools', 'window', '-sm',
+                          '-w', str(window), 
+                          '-a', bed1, 
+                          '-b', bed2], 
+                         capture_output=True, shell=False)
+
+    assert out.returncode==0, "bedtools window run failed, check input files"
+
+    # memory file-handle to pass output to pandas without writing to disk
+    out_handle = StringIO(out.stdout.decode())
+    out = pd.read_csv(out_handle, delimiter='\t', header=None, dtype={0: str})
+    
+    return(out)
+
+def overlap(coords1, coords2):
+    """
+    Calculates number of points overlapping when the input are two intervals as tuples.
+    Input must be coords1=(int, int),  coords2=(int, int).
+    """
+    return(len(np.intersect1d(np.arange(coords1[0], coords1[1]), np.arange(coords2[0], coords2[1]))))
+
+
+print('-------------------------------------------------')
+## sum expression levels for predicted sites that overlap with each other given the window size
+## assuming the prediction bed file has exactly 6 columns in the correct order, no header!
+
+# find sites that overlap within the predict set itself when the window is extended
+out = bedtools_window(f_PD, f_PD, window)
+
+# for logging
+len_dup = len(out)
+
+not_unq_mask = out.duplicated([0, 1, 2, 5], keep=False)
+
+MERGED=False
+if np.sum(not_unq_mask) > 0: # otherwise there were no overlapping sites
+    MERGED=True
+    
+    not_unq = out[not_unq_mask]
+
+    # remove non-unique sites
+    out.drop(out.index[not_unq_mask], inplace=True, axis=0)
+    out.drop([6, 7, 8, 9, 10, 11], inplace=True, axis=1)
+
+    # collapse non-unique sites, sum up expression, add back to bed file
+    unq = np.unique([(i, j, k, l) for i, j, k, l in zip(not_unq[0], not_unq[1], not_unq[2], not_unq[5])], axis=0)
+
+    for i in unq:
+        non_unq_context = not_unq[(not_unq[0]==i[0]) & (not_unq[1]==int(i[1])) & (not_unq[2]==int(i[2])) & (not_unq[5]==i[3])]
+
+        merged_start = min(list(non_unq_context[7]))
+        merged_end = max(list(non_unq_context[8]))
+        merged_exp = sum(non_unq_context[10])
+
+        out = out.append({0: i[0], 1: merged_start, 2: merged_end, 3: list(non_unq_context[3])[0], 4: merged_exp, 5: i[3]}, ignore_index=True)
+
+    # de-duplicate, sort
+    out.drop_duplicates([0, 1, 2, 5], inplace=True)
```

Comment:
Duplicates are removed but I'm not sure if it solves the problem if multiple different merged sites are created. For example when we have sites at locations: 4, 8, 12, 16 with window of size 6, we will end up with several overlapping sites, right?

Reply:
I'm not 100% sure I understand the question. Can you describe the overlap scenario you think can cause a problem? 

The de-duplication is because artificial duplicates are added by the for loop above. I found this was easier than checking every time in the loop if weighting on one PAS had already been performed.

Reply:
bedtools window will only find extended sites from one file overlapping with files from the other file. But I don't think it constructs single-linkage clusters of overlapping sites. So yes, I think we would end up with multiple overlapping sites from the predictions file.

Going back to the original issue, to try to deal with some degree of noise in 3' end processing, I would suggest to not merge the predictions, but take them face-value as individual sites. One can extend them (but a small number of nucleotides) when searching for overlaps with the ground truth. Then, the following scenarios can occur:
1. extended predicted site does not overlap with any ground truth site - the ground truth site expression can be set to 0 (the rest of the fields can be set from the predicted site)
2. extended predicted site overlaps with one ground truth site - sites are matched with each other
3. extended ground truth site overlap with multiple ground truth sites - if included, the expression of this site should be split between the ground truth sites (to make it clear the name can acquire a suffix and the genomic interval can be split, but depending on the functions used, it may be OK to have duplicate name and coordinates). How to split? Simplest would be in proportion to the overlap of the genomic intervals (as currently done in the code).

Reply:
In a scenario with PAS at: 10, 20, 30 and 10 nt window we'd end up with merged PAS like so: 10-20, 10-30 and 20-30. If each of individual sites had TPM=20, then cummulative TPM of merged sites would be: 40, 60, 40. Now assuming GT sites are at: 5 and 35, then mapping would be: 

- 10-20 site mapped to GT-5 with weight=1, 
- 10-30 site mapped to GT-5 with weight=0.5, 
- 10-30 site mapped to GT-35 with weight=0.5,
- 20-30 site mapped to GT-35 with weight=0.5

The cummulative TPM of predicted sites mapped to GT-5 would be then 40 * 1 + 60 * 0.5 = 70, same for GT-35. When in reality it should be closer to 30 for each (20 from the closest predicted site, and maybe 10 from the site between them).

I don't know how likely a similar scenario is, but I think in some edge cases, depending on tools output or window size, it can be problematic.

(If I got something really wrong here, I'm sorry for the essay)